### DISCUSSION TITLE
Vectorized annotations/glyphs

### DISCUSSION THREAD

**Discussion by ianthomas23** [bokeh/bokeh#12487] [Internals and design (Q&A)]
There are a few historic issues about vectorized annotations (#9955, #11015) that I don't want to append to so I am starting a new discussion here.

I have a use case for vectorized annotations which are all based on identifying vertical lines (e.g. `VSpan`) and vertical blocks (e.g. `BoxAnnotation` or `Band`) in timeseries plots, i.e. line plots where the `x` values are monotonically increasing. You can add individual annotations now, but I would like to be able to add a large number of them, let's say a million (in a very long timeseries that you only display a part of at any one time).

There is the possibility of vectorizing the specific annotation classes of interest, but the requirement for such large numbers means that they need to the glyphs rather than annotations to have decent performance. If they have to be glyphs, what are the features missing from glyphs that are needed for this annotation-like use case?

1. Support infinite extent, horizontally and/or vertically.
2. Allow auto-ranging to be configurable: on and off, one or both dimensions.

In the short term the way forward seems to be to implement the missing features listed above for glyphs, then create two new glyph classes. Longer term there is the question of what to do with annotations and glyphs. Do they remain separate class hierarchies, or should they be combined into one with support for both possible forms of behaviour which is configurable by the user as required? If they are combined into a single class hierarchy does that exclude specific awkward annotations like LaTeX rendering ones that are potentially quite slow to render if vectorized?

**Comment by bryevdv** [comment]
I don't have much to add here beside history. I will explain how things came to be how they currently are, which may be helpful to inform future work. 

* Early on, we added spatial indexing support for efficient hit-testing (specifically the rbush library)
* Spatial indexing has to occur in data-space, otherwise you'd need to re-index every time there was a pan or zoom

However:

* In the beginning, glyphs could be configured with *screen-space* coordinates. Screen space coordinates are often useful for "annotations"

And that's causes the tension. It became impossible to manage glyphs that could be configured with both data- *or* screen-space coordinates [1] while making efficient use of the spatial index at the same time. 

So things were split:

* Glyphs had spatial indexing, but only supported data-space coordinates
* Annotations could support screen-space coordinates, but did not participate in hit-testing (no spatial index)

I am not at all opposed to moving toward some sort of unification, but whatever the proposal, it needs to square that circle. That might be by having glyphs that toggle off spatial indexing if screen-space coordinates are used. That maybe doesn't sound too bad until you consider the data/screen space could be changed dynamically at any time.  Or maybe we ditch the spatial index altogether for something else that can work well-enough in all cases. 

I also think, even if there is a unification of implementations, that there is conceptual value in having dedicated API for "annotations" because those are common concepts in the dataviz world, and we should map to user expectations whenever possible. 

[1] or even *both!* earliest glyphs could have "data" for x and "screen" for y

**Comment by bryevdv** [comment]
Specifically regarding the current PR #12677 I have a number of concerns:

* `HSpan` and `VSpan` are duplicative of both the name and functionality `Span`
* `HBand` and `VBand` are duplicative of the name but *not the functionality* of `Band` (arguably worse).

Adding new things has knock on effects for documentation and gallery, and associated costs if there are churn. I am opposed to adding anything new until there is a documented and agreed plan forward. 

Here are some things I would advocate FOR:

* Generally speaking, annotations should support hit testing
* Annotations should support being added to legends

And AGAINST:

* Categorizing anything and everything a "glyph". 

There is continued instructive value in taxonomies, even if the implementation details converge to be identical.

So, examples of questions:

*  Do `HSpan` and `HBand` need to be distinct? Can we have one model `HSpan` that supports spanning a given direction, including optional filling/hatching? If so we could deprecate `Span` and there would be no conflict with `Band` 
* Can we merge the features of `Text` and `LabelSet`, and deprecate one of them?  (I'd say the name "label" has an important weight and significance in the space of dataviz and that would be the better one to keep)
* We still need a very convenient way to add *single* labels, how does that fit in?
* What about the proliferation of "HTMLxxxx" annotations?
* In general: what is a *plan* that can guide future work? What is the concrete criteria for classifying things as annotations? 

i.e. do we want to say: "Annotations with data sources must support vectorization and hit-testing". Is there a meaningful name to distinguish vectorized annotations with data sources (like spans or labels) from those without? (i.e. a legend, or colorbar)? I think more typically the former might just be "annotation" and the latter be a "guide". We never previously defined a categoriy for guides, but we could, as a matter of grouping and documentation. 

And I am sure others. I don't want to "answer" these questions ad-hoc, one at a time, since: *that is how we got to where things are currently*.

**Reply by bryevdv** [reply]
More concretely: When I say "what is a plan" I more or less mean: What is desired end-state for the contents of `bokeh.models.annotations` (which matters for documentation sectioning) vs other modules? What exists now that we will replace or deprecate? A full and complete current census, with a listing of what goes where and why, and some guidelines for to apply for future development.

**Comment by bryevdv** [comment]
cc @bokeh/dev this is briefer than I had hoped to write down, but I am under a bti of a time crunch. I hope this gets the high-level idea across

### Past / Future

1. Early on, implementation and docs were aligned with the notion that annotations were *things*, represented by specific models
1. A technical consideration (screen-space coords and hit-testing) reinforced this dichotomy of "glyphs" vs "annotations". 

These things made seemed reasonable at the time, but Bokeh was much smaller than and some of the assumptions no longer line up with with users actually want. It's clear that there is a demand for annotations to participate in legends, hit-testing, etc. So I think a simple solution is to re-focus along these definitions:

1. Glyphs are models that draw that can be driven by a CDS
2. Annotation is an *action*, not a *thing*

### Glyph = CDS + Draw

I would keep things this simple. It's easy to explain and implement. 

More specifically: 

* I would *not* require glyphs to participate in hit-testing at all, or to use `RBush` if they do (e.g. some could just offer "brute force" hit testing). 

* I would *not* require glyphs have any particular affordances for data vs screen coordinates or a mix thereof

What I *do* think needs to happen is that we develop some way to automatically keep track of which glyphs offer what, and include that information the docs. E.g. right now there is just a manual wiki page for hit-testing that is often out-of-date. 

### Annotation is an action

By his I mean that the docs should re-focus away from, e.g *"Here is a `Span` annotation object"* to *"Here is how you can annotate your plot with linear spans (using glyphs like `Span` and `Slope`)"* I think decoupling the *act* of annotation things from distinct models actually affords more flexibility in how we can document things. 

### Plan

If folks like the general shape of this plan, there's still details to figure out:

* There are a few things that *aren't* driven by CDS&mdash;where do these go, how do we describe them, etc? Can they be adapted to use CDS (e.g. arrows are a good candidate)
* What, specifically, needs to be deprecated, and on what timeline? (e.g. label vs labelset vs text...)
* Is `class Annotation(Glyph)` the best and least disruptive path?

What would be helpful is for someone(s) to sit down and do an actual census of all the current annotations and document exactly how far or close they already are from "being a glyph".

**Comment by mattpap** [comment]
Related discussions and issues: https://github.com/bokeh/bokeh/issues/11012, https://github.com/bokeh/bokeh/issues/11015, https://github.com/bokeh/bokeh/issues/10644.
